File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 118667. 두 큐 합 같게 만들기/kotlin/Solution.kt
class Solution {
    fun solution(queue1: IntArray, queue2: IntArray): Int {
        val size = queue1.size
        val q1 = Queue(2*size)
        val q2 = Queue(2*size)
        for (e in queue1) q1.eq(e)
        for (e in queue2) q2.eq(e)
        val twoQueue = TwoQueue(q1, q2)
        if (twoQueue.hasSameSum()) return 0
        for (answer in 1..3*size) {
            twoQueue.bigToSmall()
            if (twoQueue.hasSameSum()) return answer
        }
        return - 1
    }
}

class Queue(cap: Int) {

    private val c = cap + 1
    private val arr = IntArray(c)
    private var front = 0
    private var rear = 0

    var sum = 0L
        private set

    fun eq(e: Int) {
        arr[((rear + 1 + c) % c).also { rear = it }] = e
        sum += e
    }

    fun dq(): Int {
        return arr[((front + 1 + c) % c).also { sum -= arr[it]; front = it }]
    }
}

class TwoQueue(
    private val q1: Queue,
    private val q2: Queue,
) {

    fun hasSameSum() : Boolean = q1.sum == q2.sum

    fun bigToSmall() {
        if (q1.sum > q2.sum) q2.eq(q1.dq())
        else if (q1.sum < q2.sum) q1.eq(q2.dq())
    }

}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 118667. 두 큐 합 같게 만들기/python/solution.py
from collections import deque


def solution(queue1, queue2):
    sum1 = sum(queue1)
    sum2 = sum(queue2)

    size = len(queue1)

    queue1 = deque(queue1)
    queue2 = deque(queue2)

    for answer in range(size * 4):
        if sum1 > sum2:
            left = queue1.popleft()
            queue2.append(left)
            sum1 -= left
            sum2 += left
        elif sum1 < sum2:
            left = queue2.popleft()
            queue1.append(left)
            sum2 -= left
            sum1 += left
        else:
            return answer
    return -1
